src/MongoDB.Analyzer/Core/Builders/BuildersResolveVariables.cs

```diff
@@ -0,0 +1,445 @@
+            AssignmentExpressionSyntax  @assignmentExpression => @assignmentExpression.Right,
+            VariableDeclaratorSyntax @variableDeclarator => @variableDeclarator.Initializer.Value,
+            _ => syntaxNode
+        };
+
+        if (rightHandSide is IdentifierNameSyntax identifierNameSyntax)
+        {
+            canEvaluate = variableValues.TryGetValueSafe(identifierNameSyntax.Identifier.ValueText, out var rightVariableValue);
+
+            if (canEvaluate)
+            {
+                syntaxNodeToVariableValueMap ??= new Dictionary<SyntaxNode, VariableValue>();
+                syntaxNodeToVariableValueMap.Add(identifierNameSyntax, rightVariableValue);
+
+                if (displayDiagnostics)
+                {
+                    rightVariableValue.Locations.Add(identifierNameSyntax.GetLocation());
+                }
+            }
+        }
+        else if (rightHandSide is InvocationExpressionSyntax invocationExpression)
+        {
+            canEvaluate = builderToAnalysisContext.TryGetValueSafe(invocationExpression, out var expressionAnalysisContext);
+
+            if (canEvaluate)
+            {
+                syntaxNodeToVariableValueMap ??= new Dictionary<SyntaxNode, VariableValue>();
+                var rightVariableValue = new VariableValue(expressionAnalysisContext, expressionAnalysisContext.Node.Locations.ToList());
+                syntaxNodeToVariableValueMap.Add(invocationExpression, rightVariableValue);
+            }
+        }
+        else if (rightHandSide is BinaryExpressionSyntax binaryExpressionSyntax)
+        {
+            canEvaluate = builderToAnalysisContext.TryGetValueSafe(binaryExpressionSyntax, out var expressionAnalysisContext);
+
+            if (!canEvaluate)
+            {
+                var leftEvaluationResult = PreEvaluate(binaryExpressionSyntax.Left, contextInformation, syntaxNodeToVariableValueMap : null, displayDiagnostics);
+
+                if (leftEvaluationResult.CanEvaluate)
+                {
+                    var rightEvaluationResult = PreEvaluate(binaryExpressionSyntax.Right, contextInformation, syntaxNodeToVariableValueMap : leftEvaluationResult.SyntaxNodeToVariableValueMap, displayDiagnostics);
+                    canEvaluate = leftEvaluationResult.CanEvaluate && rightEvaluationResult.CanEvaluate;
+                    syntaxNodeToVariableValueMap = rightEvaluationResult.SyntaxNodeToVariableValueMap;
+                }
+            }
+            else
+            {
+                syntaxNodeToVariableValueMap ??= new Dictionary<SyntaxNode, VariableValue>();
+                var rightVariableValue = new VariableValue(expressionAnalysisContext, expressionAnalysisContext.Node.Locations.ToList());
+                syntaxNodeToVariableValueMap.Add(binaryExpressionSyntax, rightVariableValue);
+            }
+        }
+
+        if (isCompoundExpression && canEvaluate)
```

Comment:
No need for dedicated `isCompoundExpression` var.

`if (canEvaluate and IsCompoundExpression(...))`

Reply:
Done